{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard library imports\n",
    "import sys\n",
    "import asyncio\n",
    "import aiohttp\n",
    "import aiofiles\n",
    "\n",
    "# Related third party imports\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "# Local application/library specific imports.\n",
    "# paradoxcell\n",
    "sys.path.append('/10Gdata/sjt')\n",
    "import paradoxcell as pc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>run</th>\n",
       "      <th>characteristic_key</th>\n",
       "      <th>characteristic_value</th>\n",
       "      <th>etc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E-CURD-1</td>\n",
       "      <td>ERR274309</td>\n",
       "      <td>age</td>\n",
       "      <td>not available</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E-CURD-1</td>\n",
       "      <td>ERR274309</td>\n",
       "      <td>developmental stage</td>\n",
       "      <td>LP.04 four leaves visible stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PO_0007115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E-CURD-1</td>\n",
       "      <td>ERR274309</td>\n",
       "      <td>ecotype</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0005148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E-CURD-1</td>\n",
       "      <td>ERR274309</td>\n",
       "      <td>genotype</td>\n",
       "      <td>wild type</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E-CURD-1</td>\n",
       "      <td>ERR274309</td>\n",
       "      <td>growth condition</td>\n",
       "      <td>10/14 h light/dark period and a light intensit...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24009</th>\n",
       "      <td>E-MTAB-964</td>\n",
       "      <td>ERR163136</td>\n",
       "      <td>cultivar</td>\n",
       "      <td>Zheng58 inbred line</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24010</th>\n",
       "      <td>E-MTAB-964</td>\n",
       "      <td>ERR163136</td>\n",
       "      <td>developmental stage</td>\n",
       "      <td>flowering stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PO_0007616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24011</th>\n",
       "      <td>E-MTAB-964</td>\n",
       "      <td>ERR163136</td>\n",
       "      <td>organism</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_4577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24012</th>\n",
       "      <td>E-MTAB-964</td>\n",
       "      <td>ERR163136</td>\n",
       "      <td>organism part</td>\n",
       "      <td>immature style</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24013</th>\n",
       "      <td>E-MTAB-964</td>\n",
       "      <td>ERR163136</td>\n",
       "      <td>organism part</td>\n",
       "      <td>immature style</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23908 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            study        run   characteristic_key  \\\n",
       "0        E-CURD-1  ERR274309                  age   \n",
       "1        E-CURD-1  ERR274309  developmental stage   \n",
       "2        E-CURD-1  ERR274309              ecotype   \n",
       "3        E-CURD-1  ERR274309             genotype   \n",
       "4        E-CURD-1  ERR274309     growth condition   \n",
       "...           ...        ...                  ...   \n",
       "24009  E-MTAB-964  ERR163136             cultivar   \n",
       "24010  E-MTAB-964  ERR163136  developmental stage   \n",
       "24011  E-MTAB-964  ERR163136             organism   \n",
       "24012  E-MTAB-964  ERR163136        organism part   \n",
       "24013  E-MTAB-964  ERR163136        organism part   \n",
       "\n",
       "                                    characteristic_value  \\\n",
       "0                                          not available   \n",
       "1                        LP.04 four leaves visible stage   \n",
       "2                                                  Col-0   \n",
       "3                                              wild type   \n",
       "4      10/14 h light/dark period and a light intensit...   \n",
       "...                                                  ...   \n",
       "24009                                Zheng58 inbred line   \n",
       "24010                                    flowering stage   \n",
       "24011                                           Zea mays   \n",
       "24012                                     immature style   \n",
       "24013                                     immature style   \n",
       "\n",
       "                                                 etc  \n",
       "0                                                NaN  \n",
       "1          http://purl.obolibrary.org/obo/PO_0007115  \n",
       "2               http://www.ebi.ac.uk/efo/EFO_0005148  \n",
       "3                                                NaN  \n",
       "4                                                NaN  \n",
       "...                                              ...  \n",
       "24009                                            NaN  \n",
       "24010      http://purl.obolibrary.org/obo/PO_0007616  \n",
       "24011  http://purl.obolibrary.org/obo/NCBITaxon_4577  \n",
       "24012                                            NaN  \n",
       "24013                                            NaN  \n",
       "\n",
       "[23908 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data download from gui\n",
    "path_ea_directroy = '/10Gdata/sjt/data/database/expression_atlas/condensed_plant_baseline_107'\n",
    "\n",
    "data_ea_condensed_sdrf_where_plant_baseline = pc.DataFile(\n",
    "    path_ea_directroy,\n",
    "    'condensed-sdrf.tsv',\n",
    "    'iso-8859-1',\n",
    "    '\\t'\n",
    "    )\n",
    "\n",
    "df_ea_condensed_sdrf_where_plant_baseline = data_ea_condensed_sdrf_where_plant_baseline.get_df_open()\n",
    "\n",
    "df_ea_condensed_sdrf_where_plant_baseline.replace('organism [part','organism part',inplace=True)\n",
    "df_ea_condensed_sdrf_where_plant_baseline.replace('wild type genotype','wild type',inplace=True)\n",
    "df_ea_condensed_sdrf_where_plant_baseline.drop_duplicates(inplace=True)\n",
    "df_ea_condensed_sdrf_where_plant_baseline.drop([3],axis=1,inplace=True)\n",
    "\n",
    "df_ea_condensed_sdrf_where_plant_baseline.columns = ['study','run','characteristic_key','characteristic_value','etc']\n",
    "df_ea_condensed_sdrf_where_plant_baseline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3405/3405 [00:19<00:00, 171.79it/s]\n"
     ]
    }
   ],
   "source": [
    "columns_df_ea_condensed_sdrf_where_run_characteristics = list(set(df_ea_condensed_sdrf_where_plant_baseline['characteristic_key']))\n",
    "columns_df_ea_condensed_sdrf_where_run_characteristics.append('run')\n",
    "columns_df_ea_condensed_sdrf_where_run_characteristics = sorted(columns_df_ea_condensed_sdrf_where_run_characteristics)\n",
    "\n",
    "list_run_characteristic = list()\n",
    "for i in tqdm(list(set(df_ea_condensed_sdrf_where_plant_baseline['run']))):\n",
    "    tmp_dict_run_characteristic = dict()\n",
    "    tmp_dict_run_characteristic['run'] = i\n",
    "    for index,row in df_ea_condensed_sdrf_where_plant_baseline.set_index('run').loc[i][['characteristic_key','characteristic_value']].iterrows():\n",
    "        tmp_dict_run_characteristic[row['characteristic_key']] = row['characteristic_value']\n",
    "    list_run_characteristic.append(tmp_dict_run_characteristic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3405/3405 [00:08<00:00, 409.06it/s]\n"
     ]
    }
   ],
   "source": [
    "df_ea_condensed_sdrf_where_run_characteristics = pd.DataFrame(columns=columns_df_ea_condensed_sdrf_where_run_characteristics).set_index('run')\n",
    "\n",
    "for i in tqdm(list_run_characteristic):\n",
    "    df = pd.DataFrame(i,index=[i['run']])\n",
    "    df_ea_condensed_sdrf_where_run_characteristics = pd.concat([df_ea_condensed_sdrf_where_run_characteristics,df])\n",
    "\n",
    "df_ea_condensed_sdrf_where_run_characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stringid_uniprotid_ensemblgenomeid_og_for_eggnog33090 = pd.read_csv('/10Gdata/sjt/data/dataframe/df_stringid_uniprotid_ensemblgenomeid_og_for_eggnog33090.tsv',sep='\\t')\n",
    "dic_ensemblgenomeid_og = dict()\n",
    "\n",
    "for key,value in zip(df_stringid_uniprotid_ensemblgenomeid_og_for_eggnog33090['ensemblgenomeid'],df_stringid_uniprotid_ensemblgenomeid_og_for_eggnog33090['og']):\n",
    "    dic_ensemblgenomeid_og[key] = value\n",
    "\n",
    "def get_df_median(df):\n",
    "    col = df.columns\n",
    "    for i in df.columns[2:]:\n",
    "        df[i] = df[i].apply(lambda x : np.median(np.array([float(v) for v in x.split(',')])))\n",
    "    return df\n",
    "\n",
    "def get_og_column(x):\n",
    "    try:\n",
    "        return dic_ensemblgenomeid_og[x]\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "def get_df_dataset_tpm(configurationfile,tpmsfile):\n",
    "    tree = ET.parse(configurationfile)\n",
    "    root = tree.getroot()\n",
    "    dic_id_assays = dict()\n",
    "    for assay_group in root[0][0].findall('assay_group'):\n",
    "        id = assay_group.get('id')\n",
    "        assays = list()\n",
    "        for assay in assay_group:\n",
    "            assays.append(assay.text)\n",
    "        dic_id_assays[id] = \",\".join(assays)\n",
    "\n",
    "    df = pd.read_csv(tpmsfile,sep='\\t')\n",
    "    df = get_df_median(df)\n",
    "    df.rename(columns=dic_id_assays,inplace=True)\n",
    "    df['og'] = df['GeneID'].apply(lambda x : get_og_column(x))\n",
    "    df = df.dropna(subset=['og']).set_index(['og']).groupby(level=0).sum()\n",
    "    return df.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ea_plant_baseline_configurationfiles = pc.DataFile(\n",
    "    '/10Gdata/sjt/data/database/expression_atlas/condensed_plant_baseline_107',\n",
    "    'configuration.xml',\n",
    "    'iso-8859-1',\n",
    "    '\\t'\n",
    ")\n",
    "\n",
    "ea_plant_baseline_tpmfiles = pc.DataFile(\n",
    "    '/10Gdata/sjt/data/database/expression_atlas/condensed_plant_baseline_107',\n",
    "    'tpms.tsv',\n",
    "    'iso-8859-1',\n",
    "    '\\t'\n",
    ")\n",
    "\n",
    "configurationfiles = sorted(ea_plant_baseline_configurationfiles.get_path())\n",
    "tpmfiles = sorted(ea_plant_baseline_tpmfiles.get_path())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dataset_tpm = pd.DataFrame()\n",
    "\n",
    "for configurationfile,tpmfile in tqdm(zip(configurationfiles,tpmfiles)):\n",
    "    df = get_df_dataset_tpm(configurationfile,tpmfile)\n",
    "    df_dataset_tpm = pd.merge(df_dataset_tpm,df,left_index=True,right_index=True,how='outer')\n",
    "\n",
    "df_dataset_tpm.drop([i for i in df_dataset_tpm.columns if i.startswith('Gene Name')],axis=1,inplace=True)\n",
    "\n",
    "df_dataset_tpm"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
